feat(allocator): add `Vec::into_boxed_slice`

DonIsaac opened the pull request:
Note that this PR does not implement the inverse operation (`Box::to_vec` for `[T]`).

DonIsaac commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6195?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6195** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6195?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#6194** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6194?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_allocator/src/vec.rs:
````diff
@@ -101,6 +101,33 @@ impl<'alloc, T> Vec<'alloc, T> {
+    ///
+    /// assert_eq!(&*b, &[1, 2, 3]);
+    /// ```
+    pub fn into_boxed_slice(self) -> Box<'alloc, [T]> {
+        let b = self.0.into_boxed_slice();
+        let (ptr, _) = allocator_api2::boxed::Box::into_non_null(b);
+        // SAFETY: this is effectively a transmute from allocator_api2's box to
+        // our own representation. Ptr is non-null and points to initialized, well-aligned memory.
+        unsafe { Box::from_non_null(ptr) }
+    }
````
Hmm. I'm not sure about this.

I'm not that familiar with `allocator_api2`'s implementation of `into_boxed_slice`, but it sounds like it can reallocate (copy all the data), which we definitely don't want.

That makes sense for `allocator_api2` as it's generic over any kind of allocator, but in our case we're using a bump allocator which cannot recover excess capacity, so it's pointless to copy the data.

I don't know if `bumpalo::Bump` would automatically ignore the request to reallocate or not. Would you like to look into that?

Alternatively, we could get the pointer and length of the `Vec` and construct a `NonNull<[T]>` manually. That's definitely do what we want. (Note: would need to wrap the `Vec` in a `ManuallyDrop` first, as `allocator_api2`'s `Vec` is `Drop`)

This all requires quite a bit of care. I'm not completely up to speed on the safety constraints. Sometimes there are non-obvious subtleties to avoiding UB, and I usually end up having to spend a few hours boning up on arcane details before I'm sure (well at least reasonably sure!) that I've not committed some UB crime!

Do you have a concrete use case for wanting a boxed slice to data in the arena?

overlookmotel replied:
By the way, we're planning to replace this `Vec` implementation with a hand-written one with length and capacity fields as `u32`s, to bring size of `Vec` down from 32 bytes to 24 bytes (https://github.com/oxc-project/backlog/issues/18). Dunqing was going to look at doing that fairly soon.

So at that point, we wouldn't be able to use `allocator_api2`'s `into_boxed_slice` anyway.

DonIsaac replied:
https://github.com/zakarumych/allocator-api2/blob/f937b9650e10137f782698bd4eb4f8b36f1817a7/src/stable/vec/mod.rs#L1046

I don't think it creates a new allocation. I want this because it cuts the size of a dynamically sized vec in half (32 -> 16 bytes) at the cost of not being re-sizeable.  I'm working on a prototype where this is useful.

overlookmotel replied:
Actually, this works:

```rs
pub fn into_boxed_slice(self) -> Box<'alloc, [T]> {
    let slice = self.0.leak();
    let ptr = NonNull::from(slice);
    // SAFETY: `ptr` points to a valid slice `[T]`.
    // Lifetime of returned `Box<'alloc, [T]>` is same as lifetime of consumed `Vec<'alloc, T>`,
    // so data in the `Box` must be valid for its lifetime.
    unsafe { Box::from_non_null(ptr) }
}
```

Then you can remove the comment about "If the vector has excess capacity, its items will be moved into a newly-allocated buffer with exactly the right capacity."

overlookmotel replied:
I've implemented this change in an extra commit.

overlookmotel commented on crates/oxc_allocator/src/boxed.rs:
```diff
@@ -76,6 +76,18 @@ impl<'alloc, T> Box<'alloc, T> {
+impl<'alloc, T: ?Sized> Box<'alloc, T> {
+    /// Create a [`Box`] from a raw pointer to a value in the arena.
+    ///
+    /// # SAFETY
+    /// It is only safe to create a `Box` this way with pointers taken out of
+    /// another `Box` in the same arena.
```
This comment should be updated. It doesn't require that the pointer is from *another `Box`*. If that was true, then `Vec:: into_boxed_slice` which you're adding here would violate the contract, as it's not a `Box`!

The requirement is that pointer needs to point to data in same arena (could be from a `Vec`, a `String`, anything else).

Actually, not even that is quite right. We *want* it to be date from the same arena, but the safety constraint I think is  a bit looser - it can point to *any* data as long as that data outlives `'alloc`.

There's also a safety requirement about no aliasing.

This may seem like nitpicking, but it's vitally important that documentation for unsafe methods is accurate about what the contract is, otherwise it's impossible to verify the method is being used correctly.

overlookmotel replied:
I've implemented this change in an extra commit.

DonIsaac commented:
@overlookmotel does it make sense to just abandon this PR?

overlookmotel commented:
No, I think we can do it. I just think better to use this version:

```rs
pub fn into_boxed_slice(self) -> Box<'alloc, [T]> {
    let slice = self.0.leak();
    let ptr = NonNull::from(slice);
    // SAFETY: `ptr` points to a valid slice `[T]`.
    // Lifetime of returned `Box<'alloc, [T]>` is same as lifetime of consumed `Vec<'alloc, T>`,
    // so data in the `Box` will be valid for its lifetime.
    unsafe { Box::from_non_null(ptr) }
}
```

It's safe, and won't copy memory.

overlookmotel commented:
@DonIsaac Hope you don't mind that I've stepped in to finish this off. Just wanted to close a few open PRs which are close to the finish line. Unless you have any objections, I plan to merge this with the additional commits I've added.

Boshen commented:
### Merge activity

* **Oct 12, 12:29 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

DonIsaac committed:
feat(allocator): add `Vec::into_boxed_slice` (#6195)

Note that this PR does not implement the inverse operation (`Box::to_vec` for `[T]`).
